README.md

```diff
@@ -1,2 +1,65 @@
-# message-queue-toolkit
-Useful utilities, interfaces and base classes for message queue handling
+# message-queue-toolkit ✉️
+Useful utilities, interfaces and base classes for message queue handling.
+
+## Overview
+
+This is an abstraction to switch between different queue systems without having to implement your own deserialization, error handling, etc. The library provides utilities, interfaces and base classes to build the support for any queue system you may need in your service and already implements support for the following:
+
+* AMQP (Advanced Message Queuing Protocol)
+* SQS (Simple Queue Service)
+* SNS (Simple Notification Service)
+
+## Basic Usage
+
+### Publishers
+
+Build a publisher to send messages to a queue or a topic in a messaging system by creating a class that:
+
+* Extends the abstract publisher class provided for the messaging system of choice
+* Defines a queue name
+* Defines a message schema
+* Configures the queue
+
+> **_NOTE:_**  See [SqsPermissionPublisher.ts](./packages/sqs/test/publishers/SqsPermissionPublisher.ts) for a practical example.
+
+#### Consumers
+
+Build a recipient of messages from a queue or a topic in a messaging system by creating a class that:
+
+* Extends the abstract consumer class provided for the messaging system of choice
+* Defines a queue name
+* Defines a message schema
+* Configures the queue
+* Overrides the `processMessage()` method
+
+> **_NOTE:_**  See [SqsPermissionConsumer.ts](./packages/sqs/test/consumers/SqsPermissionConsumer.ts) for a practical example.
+
+## Error Handling
+
+`processMessage()` return response is `Either<'retryLater', 'success'>`, meaning that it could be an either one value or the other. In case of `retryLater`, the abstract consumer is instructed to requeue the message. Otherwise, in case of success, the message is finally removed from the queue. In case of any error, the abstract consumer will also requeue the message. When overriding the `processMessage()` method, you should leverage the possible types to process the message as you need.
+
+## Schema Validation and Deserialization
+
+Message deserialization is done within the abstract consumer _before_ processing the message. Deserialization is done for a message, given a message schema type. We use `zod` to define the message schema type.
+
+If
+* The message before deserialization is `null`
+* Deserialization returns a `MessageInvalidFormatError` error for a syntax error
```

Comment:
In this context specific errors are an implementation detail, they are never propagated to the end user.

They can be mentioned if you choose to explain that you can define and pass custom `ErrorResolver` - if either of these errors are returned during error resolution, AbstractConsumer will nack the message.